// @vibecodeapp/webapp/dist/plugin.js
// src/vite-plugin.ts
import path from "path";
var VIRTUAL_MODULE_ID = "\0vibecode-jsx-source/jsx-dev-runtime";
var JSX_DEV_RUNTIME_ID = "react/jsx-dev-runtime";
function vibecodePlugin() {
  const overrides = /* @__PURE__ */ new Map();
  let server = null;
  let projectRoot = "";
  const normalizeFilePath = (filePath) => {
    if (path.isAbsolute(filePath)) return filePath;
    return path.resolve(projectRoot, filePath);
  };
  const invalidateModule = async (filePath) => {
    if (!server) return;
    const normalizedPath = normalizeFilePath(filePath);
    const module = server.moduleGraph.getModuleById(normalizedPath);
    if (module) {
      server.moduleGraph.invalidateModule(module);
      server.ws.send({
        type: "update",
        updates: [{
          type: module.type === "css" ? "css-update" : "js-update",
          path: module.url,
          acceptedPath: module.url,
          timestamp: Date.now()
        }]
      });
    } else {
      server.ws.send({ type: "full-reload" });
    }
  };
  return {
    name: "vibecode",
    enforce: "pre",
    configResolved(config) {
      projectRoot = config.root;
    },
    configureServer(devServer) {
      server = devServer;
      devServer.ws.on("vibecode:override", async (data) => {
        const normalizedPath = normalizeFilePath(data.path);
        overrides.set(normalizedPath, data.content);
        await invalidateModule(data.path);
      });
      devServer.ws.on("vibecode:clear-override", async (data) => {
        const normalizedPath = normalizeFilePath(data.path);
        overrides.delete(normalizedPath);
        await invalidateModule(data.path);
      });
      devServer.ws.on("vibecode:clear-all-overrides", async () => {
        const paths = Array.from(overrides.keys());
        overrides.clear();
        for (const p of paths) {
          await invalidateModule(p);
        }
      });
    },
    resolveId(id, importer) {
      if (id === JSX_DEV_RUNTIME_ID && !importer?.includes("\0vibecode-jsx-source")) {
        return VIRTUAL_MODULE_ID;
      }
      return null;
    },
    load(id) {
      const override = overrides.get(id);
      if (override !== void 0) {
        return override;
      }
      const idWithoutQuery = id.split("?")[0];
      const overrideWithoutQuery = overrides.get(idWithoutQuery);
      if (overrideWithoutQuery !== void 0) {
        return overrideWithoutQuery;
      }
      if (id === VIRTUAL_MODULE_ID) {
        return `
import { Fragment, jsxDEV as _jsxDEV } from 'react/jsx-dev-runtime';

const SOURCE_KEY = Symbol.for('vibecode.__jsxSource__');

// Global map of source location -> WeakRef<Element>[]
// Allows looking up DOM elements by their source file:line:column
const sourceElementMap = new Map();
window.vibecodeSourceElementMap = sourceElementMap;

function getSourceKey(sourceInfo) {
  return \`\${sourceInfo.fileName}:\${sourceInfo.lineNumber}:\${sourceInfo.columnNumber}\`;
}

function registerElement(node, sourceInfo) {
  const key = getSourceKey(sourceInfo);
  if (!sourceElementMap.has(key)) {
    sourceElementMap.set(key, new Set());
  }
  sourceElementMap.get(key).add(new WeakRef(node));
}

function unregisterElement(node, sourceInfo) {
  const key = getSourceKey(sourceInfo);
  const refs = sourceElementMap.get(key);
  if (refs) {
    for (const ref of refs) {
      if (ref.deref() === node) {
        refs.delete(ref);
        break;
      }
    }
    if (refs.size === 0) {
      sourceElementMap.delete(key);
    }
  }
}

function getTypeName(type) {
  if (typeof type === 'string') return type;
  if (typeof type === 'function') return type.displayName || type.name || null;
  if (typeof type === 'object' && type !== null) {
    return type.displayName || type.render?.displayName || type.render?.name || null;
  }
  return null;
}

export function jsxDEV(type, props, key, isStaticChildren, source, self) {
  if (source?.fileName && typeof type !== 'string' && type !== Fragment) {
    const canHaveRef =
      type.$$typeof === Symbol.for('react.forward_ref') ||
      type.render ||
      (type.prototype && type.prototype.isReactComponent);

    if (!canHaveRef) {
      return _jsxDEV(type, props, key, isStaticChildren, source, self);
    }

    const typeName = getTypeName(type);
    const sourceInfo = {
      fileName: source.fileName,
      lineNumber: source.lineNumber ?? null,
      columnNumber: source.columnNumber ?? null,
      displayName: typeName,
      componentName: typeName,
    };

    const originalRef = props?.ref;
    const enhancedProps = {
      ...props,
      ref: (node) => {
        if (node) {
          // Only tag if not already tagged (skip on re-renders)
          if (!node[SOURCE_KEY]) {
            node[SOURCE_KEY] = sourceInfo;
            registerElement(node, sourceInfo);
          }
        }
        if (typeof originalRef === 'function') {
          originalRef(node);
        } else if (originalRef && typeof originalRef === 'object') {
          originalRef.current = node;
        }
      },
    };

    return _jsxDEV(type, enhancedProps, key, isStaticChildren, source, self);
  }
  if (source?.fileName && typeof type === 'string') {
    const sourceInfo = {
      fileName: source.fileName,
      lineNumber: source.lineNumber ?? null,
      columnNumber: source.columnNumber ?? null,
      displayName: type,
      componentName: null,
    };

    const originalRef = props?.ref;
    const enhancedProps = {
      ...props,
      ref: (node) => {
        if (node) {
          const existingSource = node[SOURCE_KEY];
          if (existingSource) {
            if (getSourceKey(existingSource) !== getSourceKey(sourceInfo)) {
              unregisterElement(node, existingSource);
              node[SOURCE_KEY] = sourceInfo;
              registerElement(node, sourceInfo);
            }
          } else {
            node[SOURCE_KEY] = sourceInfo;
            registerElement(node, sourceInfo);
          }
        }
        if (typeof originalRef === 'function') {
          originalRef(node);
        } else if (originalRef && typeof originalRef === 'object') {
          originalRef.current = node;
        }
      },
    };

    return _jsxDEV(type, enhancedProps, key, isStaticChildren, source, self);
  }

  return _jsxDEV(type, props, key, isStaticChildren, source, self);
}

export { Fragment };
`;
      }
      return null;
    },
    transformIndexHtml(html) {
      const hmrBridge = `<script type="module">
import{createHotContext}from'/@vite/client';const h=createHotContext('/@vibecode-hmr');
window.addEventListener('message',e=>{const d=e.data;
if(d?.type==='vibecode:override')h.send('vibecode:override',{path:d.path,content:d.content});
else if(d?.type==='vibecode:clear-override')h.send('vibecode:clear-override',{path:d.path});
else if(d?.type==='vibecode:clear-all-overrides')h.send('vibecode:clear-all-overrides',{});});
</script>`;
      const sdkLoader = `<script>
(function(){
  var loaded=false;
  var allowedOrigins=['https://vibecodeapp.com','https://www.vibecodeapp.com','https://vibecode.dev','http://www.vibecode.dev','http://localhost:3001'];
  var allowedSdkUrls=['https://images.composerapi.com/','https://api.vibecodeapp.com/'];
  window.addEventListener('message',function(e){
    if(!allowedOrigins.includes(e.origin))return;
    if(e.data?.type==='vibecode:sdk:load'&&!loaded){
      if(!allowedSdkUrls.some(function(u){return e.data.url.startsWith(u)}))return;
      loaded=true;
      window.__VIBECODE_CONFIG__=e.data;
      var s=document.createElement('script');
      s.src=e.data.url;
      s.onerror=function(){console.warn('[vibecode] SDK failed to load')};
      document.head.appendChild(s);
    }
  });
  if(window.parent!==window){
    window.parent.postMessage({type:'vibecode:sdk:ready-for-config'},'*');
  }
})();
</script>`;
      return html.replace("</head>", `${hmrBridge}
${sdkLoader}
</head>`);
    }
  };
}
export {
  vibecodePlugin
};
